File: src/agenticflow_cli/policy.py
"""Policy and audit helpers for AgenticFlow CLI runtime guardrails."""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Mapping

from agenticflow_cli.spec import Operation

POLICY_FILE_NAME = "policy.json"
AUDIT_LOG_FILE_NAME = "agenticflow-audit.log"
DEFAULT_POLICY_VERSION = 1
CONFIG_DIR_ENV_VAR = "AGENTICFLOW_CLI_DIR"
POLICY_FILE_ENV_VAR = "AGENTICFLOW_POLICY_FILE"
AUDIT_LOG_ENV_VAR = "AGENTICFLOW_AUDIT_LOG_FILE"


class PolicyConfigError(ValueError):
    """Raised when policy JSON cannot be loaded or written."""

    code: str
    retryable: bool
    detail: str

    def __init__(self, code: str, detail: str, retryable: bool = False) -> None:
        super().__init__(detail)
        self.code = code
        self.detail = detail
        self.retryable = retryable


@dataclass(frozen=True)
class PolicyConfig:
    spend_ceiling: float | None = None
    allowlist: tuple[str, ...] = ()
    blocklist: tuple[str, ...] = ()

    def to_dict(self) -> dict[str, Any]:
        return {
            "spend_ceiling": self.spend_ceiling,
            "allowlist": list(self.allowlist),
            "blocklist": list(self.blocklist),
        }


@dataclass(frozen=True)
class PolicyViolation:
    code: str
    detail: str
    retryable: bool = False


def _resolve_cli_dir(config_dir: Path | None = None) -> Path:
    if config_dir is not None:
        return config_dir.expanduser()
    env_dir = os.getenv(CONFIG_DIR_ENV_VAR)
    if env_dir:
        return Path(env_dir).expanduser()
    return (Path.home() / ".agenticflow").expanduser()


def policy_file_path(
    *,
    policy_file: Path | None = None,
    config_dir: Path | None = None,
) -> Path:
    if policy_file is not None:
        return policy_file.expanduser()
    env_path = os.getenv(POLICY_FILE_ENV_VAR)
    if env_path:
        return Path(env_path).expanduser()
    return _resolve_cli_dir(config_dir) / POLICY_FILE_NAME


def audit_log_path(
    *,
    audit_log: Path | None = None,
    config_dir: Path | None = None,
    policy_file: Path | None = None,
) -> Path:
    if audit_log is not None:
        return audit_log.expanduser()
    env_path = os.getenv(AUDIT_LOG_ENV_VAR)
    if env_path:
        return Path(env_path).expanduser()
    resolved_config_dir: Path | None = None
    if config_dir is not None:
        resolved_config_dir = _resolve_cli_dir(config_dir)
    elif policy_file is not None:
        resolved_config_dir = Path(policy_file).expanduser().parent
    else:
        resolved_config_dir = _resolve_cli_dir(None)
    return resolved_config_dir / AUDIT_LOG_FILE_NAME


def _coerce_spend_value(value: Any, *, field_name: str) -> float | None:
    if value is None:
        return None
    if isinstance(value, (int, float)):
        return float(value)
    if isinstance(value, str):
        try:
            return float(value)
        except ValueError as exc:
            raise PolicyConfigError(
                "policy.invalid_spend_ceiling",
                f"Invalid value for {field_name}: {value}",
            ) from exc
    raise PolicyConfigError(
        "policy.invalid_spend_ceiling",
        f"Invalid value for {field_name}: {value}",
    )


def _coerce_operations(value: Any, *, field_name: str) -> tuple[str, ...]:
    if value is None:
        return ()
    if not isinstance(value, list):
        raise PolicyConfigError(
            "policy.invalid_operation_list",
            f"{field_name} must be a JSON array of strings.",
        )
    operations: list[str] = []
    for item in value:
        if not isinstance(item, str):
            raise PolicyConfigError(
                "policy.invalid_operation_list",
                f"{field_name} must contain only strings.",
            )
        operations.append(item)
    return tuple(operations)


def _coerce_policy_payload(payload: Mapping[str, Any]) -> PolicyConfig:
    spend_ceiling = _coerce_spend_value(
        payload.get("spend_ceiling"),
        field_name="spend_ceiling",
    )
    allowlist = _coerce_operations(payload.get("allowlist"), field_name="allowlist")
    blocklist = _coerce_operations(payload.get("blocklist"), field_name="blocklist")
    return PolicyConfig(
        spend_ceiling=spend_ceiling,
        allowlist=allowlist,
        blocklist=blocklist,
    )


def load_policy(
    *,
    policy_file: Path | None = None,
    config_dir: Path | None = None,
) -> PolicyConfig:
    path = policy_file_path(policy_file=policy_file, config_dir=config_dir)
    if not path.exists():
        return PolicyConfig()
    try:
        raw_text = path.read_text(encoding="utf-8")
    except OSError as exc:
        raise PolicyConfigError(
            "policy.load_failed",
            f"Failed to read policy file: {path}",
        ) from exc

    try:
        payload = json.loads(raw_text)
    except json.JSONDecodeError as exc:
        raise PolicyConfigError(
            "policy.load_failed",
            f"Policy file is not valid JSON: {path}",
        ) from exc
    if not isinstance(payload, Mapping):
        raise PolicyConfigError(
            "policy.load_failed",
            f"Policy file payload must be an object: {path}",
        )
    try:
        return _coerce_policy_payload(payload)
    except PolicyConfigError:
        raise
    except Exception as exc:  # noqa: BLE001
        raise PolicyConfigError(
            "policy.load_failed",
            f"Invalid policy file format: {path}",
        ) from exc


def write_default_policy(
    *,
    policy_file: Path | None = None,
    config_dir: Path | None = None,
    spend_ceiling: float | None = None,
    allowlist: tuple[str, ...] = (),
    blocklist: tuple[str, ...] = (),
    force: bool = False,
) -> tuple[Path, PolicyConfig]:
    path = policy_file_path(policy_file=policy_file, config_dir=config_dir)
    if path.exists() and not force:
        raise PolicyConfigError(
            "policy.already_exists",
            f"Policy file already exists: {path}",
        )
    config = PolicyConfig(
        spend_ceiling=spend_ceiling,
        allowlist=allowlist,
        blocklist=blocklist,
    )
    path.parent.mkdir(parents=True, exist_ok=True)
    payload = {
        "version": DEFAULT_POLICY_VERSION,
        "spend_ceiling": config.spend_ceiling,
        "allowlist": list(config.allowlist),
        "blocklist": list(config.blocklist),
    }
    path.write_text(json.dumps(payload, indent=2) + "\n", encoding="utf-8")
    return path, config


def _first_float(value: Any) -> float | None:
    if isinstance(value, (int, float)):
        return float(value)
    if isinstance(value, str):
        try:
            return float(value)
        except ValueError:
            return None
    return None


def estimate_operation_cost(operation: Operation) -> float | None:
    raw = operation.raw
    if not isinstance(raw, Mapping):
        return None
    for key in (
        "x-cost",
        "x-cost-estimate",
        "x-spend",
        "x-spend-estimate",
        "estimated_cost",
    ):
        candidate = _first_float(raw.get(key))
        if candidate is not None:
            return candidate
    return None


def evaluate_policy(
    policy_config: PolicyConfig,
    operation: Operation,
    *,
    estimated_cost: float | None = None,
) -> PolicyViolation | None:
    operation_id = operation.operation_id
    if not operation_id:
        return PolicyViolation(
            code="policy.operation_missing_id",
            detail="Cannot apply policy because the operation has no operation_id.",
        )
    if operation_id in policy_config.blocklist:
        return PolicyViolation(
            code="policy.blocklisted",
            detail=f"Operation '{operation_id}' is blocklisted.",
        )
    if policy_config.allowlist and operation_id not in policy_config.allowlist:
        return PolicyViolation(
            code="policy.not_allowlisted",
            detail=(
                f"Operation '{operation_id}' is not in the allowlist policy."
            ),
        )
    if policy_config.spend_ceiling is None:
        return None

    operation_cost = estimated_cost
    if operation_cost is None:
        operation_cost = estimate_operation_cost(operation)
    if operation_cost is None:
        return None
    if operation_cost > policy_config.spend_ceiling:
        return PolicyViolation(
            code="policy.spend_ceiling_exceeded",
            detail=(
                f"Estimated cost {operation_cost:g} exceeds "
                f"spend ceiling {policy_config.spend_ceiling:g}."
            ),
        )
    return None


def audit_error_payload(
    code: str,
    detail: str,
    *,
    operation_id: str | None = None,
    retryable: bool = False,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "code": code,
        "retryable": retryable,
        "detail": detail,
    }
    if operation_id is not None:
        payload["operation_id"] = operation_id
    return payload


def write_audit_entry(
    *,
    operation_id: str,
    status: str,
    latency_ms: float,
    result_code: str,
    error: str | None = None,
    audit_path: Path | None = None,
    config_dir: Path | None = None,
    policy_file: Path | None = None,
) -> None:
    path = audit_log_path(
        audit_log=audit_path,
        config_dir=config_dir,
        policy_file=policy_file,
    )
    path.parent.mkdir(parents=True, exist_ok=True)
    payload = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "operation_id": operation_id,
        "status": status,
        "latency_ms": latency_ms,
        "result_code": result_code,
    }
    if error is not None:
        payload["error"] = error
    with path.open("a", encoding="utf-8") as handle:
        handle.write(json.dumps(payload) + "\n")

File: packages/cli/src/cli/policy.ts
/**
 * Policy and audit helpers for AgenticFlow CLI runtime guardrails.
 */

import { existsSync, mkdirSync, readFileSync, writeFileSync, appendFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { homedir } from "node:os";
import type { Operation } from "./spec.js";

export const POLICY_FILE_NAME = "policy.json";
export const AUDIT_LOG_FILE_NAME = "agenticflow-audit.log";
export const CONFIG_DIR_ENV_VAR = "AGENTICFLOW_CLI_DIR";
export const POLICY_FILE_ENV_VAR = "AGENTICFLOW_POLICY_FILE";
export const AUDIT_LOG_ENV_VAR = "AGENTICFLOW_AUDIT_LOG_FILE";

export class PolicyConfigError extends Error {
  code: string;
  retryable: boolean;
  detail: string;

  constructor(code: string, detail: string, retryable = false) {
    super(detail);
    this.name = "PolicyConfigError";
    this.code = code;
    this.detail = detail;
    this.retryable = retryable;
  }
}

export interface PolicyConfig {
  spendCeiling: number | null;
  allowlist: string[];
  blocklist: string[];
}

export interface PolicyViolation {
  code: string;
  detail: string;
  retryable: boolean;
}

function resolveCliDir(configDir?: string): string {
  if (configDir) return configDir;
  const envDir = process.env[CONFIG_DIR_ENV_VAR];
  if (envDir) return envDir;
  return resolve(homedir(), ".agenticflow");
}

export function policyFilePath(options?: {
  policyFile?: string;
  configDir?: string;
}): string {
  const envFile = process.env[POLICY_FILE_ENV_VAR];
  if (envFile) return envFile;
  if (options?.policyFile) return options.policyFile;
  return resolve(resolveCliDir(options?.configDir), POLICY_FILE_NAME);
}

export function auditLogPath(options?: {
  auditLog?: string;
  configDir?: string;
  policyFile?: string;
}): string {
  const envFile = process.env[AUDIT_LOG_ENV_VAR];
  if (envFile) return envFile;
  if (options?.auditLog) return options.auditLog;
  const policyPath = policyFilePath(options);
  return resolve(dirname(policyPath), AUDIT_LOG_FILE_NAME);
}

export function loadPolicy(options?: {
  policyFile?: string;
  configDir?: string;
}): PolicyConfig {
  const filePath = policyFilePath(options);
  if (!existsSync(filePath)) {
    return { spendCeiling: null, allowlist: [], blocklist: [] };
  }

  let raw: string;
  try {
    raw = readFileSync(filePath, "utf-8");
  } catch {
    throw new PolicyConfigError("read_error", `Unable to read policy file: ${filePath}`);
  }

  let payload: Record<string, unknown>;
  try {
    payload = JSON.parse(raw);
  } catch {
    throw new PolicyConfigError("parse_error", `Invalid JSON in policy file: ${filePath}`);
  }

  return coercePolicyPayload(payload);
}

export function writeDefaultPolicy(options?: {
  policyFile?: string;
  configDir?: string;
  spendCeiling?: number | null;
  allowlist?: string[];
  blocklist?: string[];
  force?: boolean;
}): string {
  const filePath = policyFilePath(options);
  if (existsSync(filePath) && !options?.force) {
    throw new PolicyConfigError(
      "already_exists",
      `Policy file already exists: ${filePath}. Use --force to overwrite.`,
    );
  }

  const payload = {
    version: 1,
    spend_ceiling: options?.spendCeiling ?? null,
    allowlist: options?.allowlist ?? [],
    blocklist: options?.blocklist ?? [],
  };

  mkdirSync(dirname(filePath), { recursive: true });
  writeFileSync(filePath, JSON.stringify(payload, null, 2) + "\n", "utf-8");
  return filePath;
}

export function estimateOperationCost(operation: Operation): number {
  const method = operation.method.toUpperCase();
  if (method === "GET" || method === "HEAD") return 0;
  if (method === "DELETE") return 0.1;
  return 1.0;
}

export function evaluatePolicy(
  policy: PolicyConfig,
  operation: Operation,
  options?: { estimatedCost?: number },
): PolicyViolation[] {
  const violations: PolicyViolation[] = [];

  if (policy.blocklist.length > 0 && policy.blocklist.includes(operation.operationId)) {
    violations.push({
      code: "blocked",
      detail: `Operation ${operation.operationId} is blocked by policy.`,
      retryable: false,
    });
  }

  if (policy.allowlist.length > 0 && !policy.allowlist.includes(operation.operationId)) {
    violations.push({
      code: "not_allowed",
      detail: `Operation ${operation.operationId} is not in the allowlist.`,
      retryable: false,
    });
  }

  const cost = options?.estimatedCost ?? estimateOperationCost(operation);
  if (policy.spendCeiling != null && cost > policy.spendCeiling) {
    violations.push({
      code: "spend_ceiling",
      detail: `Estimated cost ${cost} exceeds ceiling ${policy.spendCeiling}.`,
      retryable: false,
    });
  }

  return violations;
}

export function writeAuditEntry(options: {
  operationId: string;
  status: string;
  latencyMs: number;
  resultCode: string;
  error?: string;
  auditPath?: string;
  configDir?: string;
  policyFile?: string;
}): void {
  const logPath = auditLogPath({ auditLog: options.auditPath, configDir: options.configDir, policyFile: options.policyFile });
  const entry = {
    timestamp: new Date().toISOString(),
    operation_id: options.operationId,
    status: options.status,
    latency_ms: Math.round(options.latencyMs),
    result_code: options.resultCode,
    ...(options.error ? { error: options.error } : {}),
  };

  try {
    mkdirSync(dirname(logPath), { recursive: true });
    appendFileSync(logPath, JSON.stringify(entry) + "\n", "utf-8");
  } catch {
    // Audit writes are best-effort
  }
}

// --- Internal helpers ---

function coercePolicyPayload(payload: Record<string, unknown>): PolicyConfig {
  return {
    spendCeiling: coerceSpendValue(payload["spend_ceiling"]),
    allowlist: coerceOperations(payload["allowlist"]),
    blocklist: coerceOperations(payload["blocklist"]),
  };
}

function coerceSpendValue(value: unknown): number | null {
  if (value == null) return null;
  const num = Number(value);
  if (isNaN(num)) return null;
  return num;
}

function coerceOperations(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value.filter((item): item is string => typeof item === "string");
}
